core/types/system_tx.go

```diff
@@ -0,0 +1,87 @@
+package types
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/rlp"
+)
+
+type SystemTx struct {
+	From  common.Address  // pre-determined sender
+	To    *common.Address // system contract
+	Gas   uint64          // gas limit
+	Nonce uint64          // nonce
+	Data  []byte          // calldata
+}
+
+func (tx *SystemTx) txType() byte { return SystemTxType }
+
+func (tx *SystemTx) copy() TxData {
+	return &SystemTx{
+		From:  tx.From,
+		Gas:   tx.Gas,
+		Nonce: tx.Nonce,
+		To:    copyAddressPtr(tx.To),
+		Data:  common.CopyBytes(tx.Data),
+	}
+}
+
+func (tx *SystemTx) chainID() *big.Int      { return new(big.Int) }
+func (tx *SystemTx) accessList() AccessList { return nil }
+func (tx *SystemTx) data() []byte           { return tx.Data }
+func (tx *SystemTx) gas() uint64            { return tx.Gas }
+func (tx *SystemTx) gasPrice() *big.Int     { return new(big.Int) }
+func (tx *SystemTx) gasTipCap() *big.Int    { return new(big.Int) }
+func (tx *SystemTx) gasFeeCap() *big.Int    { return new(big.Int) }
+func (tx *SystemTx) value() *big.Int        { return new(big.Int) }
+func (tx *SystemTx) nonce() uint64          { return tx.Nonce }
+func (tx *SystemTx) to() *common.Address    { return tx.To }
+
+func (tx *SystemTx) rawSignatureValues() (v, r, s *big.Int) {
+	return new(big.Int), new(big.Int), new(big.Int)
+}
+
+func (tx *SystemTx) setSignatureValues(chainID, v, r, s *big.Int) {}
+
+func (tx *SystemTx) encode(b *bytes.Buffer) error {
+	return rlp.Encode(b, tx)
+}
+
+func (tx *SystemTx) decode(input []byte) error {
+	return rlp.DecodeBytes(input, tx)
+}
+
+var _ TxData = (*SystemTx)(nil)
+
+func NewOrderedSystemTxs(stxs []*SystemTx) *OrderedSystemTxs {
+	txs := make([]*Transaction, 0, len(stxs))
+	for _, stx := range stxs {
+		txs = append(txs, NewTx(stx))
+	}
+	return &OrderedSystemTxs{txs: txs}
+}
+
+type OrderedSystemTxs struct {
+	txs []*Transaction
+}
+
+func (o *OrderedSystemTxs) Peek() *Transaction {
+	if len(o.txs) > 0 {
+		return o.txs[0]
+	}
+	return nil
+}
+
+func (o *OrderedSystemTxs) Shift() {
+	if len(o.txs) > 0 {
+		o.txs = o.txs[1:]
+	}
+}
+
+func (o *OrderedSystemTxs) Pop() {
+	o.txs = nil
```

Comment:
Is this not used?

Reply:
interface documentation is not clear, but it seems pop should clear.

Reply:
Documentation states: 
```
	// Pop removes all transactions from the current account.
```

So I guess this is correct if it is a single system sender; nonetheless we can future proof it and drop from the current sender.

Reply:
ok, futured proofed it -- sort by sender as bytes, and Nonce and then Pop does what the documentation says, pops the current sender txs.

Reply:
Ok sounds good.

Maybe it helps if I provide some context:
- Originally there was no `OrderedTransactionSet`, only the single type `TransactionsByPriceAndNonce`.
- In `worker`'s `commitTransactions`, when we successfully processed a transaction, we'd use `Shift` to get the next highest paying one (either from the same account or from a different one).
- On certain errors where we could not process the transaction (e.g. nonce mismatch), we'd use `Pop` to remove all txs from this sender, since subsequent ones could not be processed anyway.
- When we introduced `L1MessageTx`, I added the `OrderedTransactionSet` interface with a new implementation. The differences is that such transactions must be processed in their `QueueIndex` order, so `Pop` is irrelevant.

This reminds me that we should remove `Nonce` from `SystemTx`, since we don't validate it anyway (see early return in `preCheck()`).

Reply:
and made it a noop in the end.